cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,627 @@
+    -- or we cannot decode it. Sadly ErrorCall can still happen, despite
+    -- using decodeFileOrFail, e.g. Data.Char.chr errors
+
+    handleDoesNotExist (Changed Nothing) $
+    handleErrorCall    (Changed Nothing) $
+          Binary.decodeFileOrFail monitorStateFile
+      >>= either (\_ -> return (Changed Nothing))
+                 checkStatusCache
+
+  where
+    -- It's also a change if the guard value has changed
+    checkStatusCache (_, cachedKey, _)
+      | currentKey /= cachedKey
+      = return (Changed Nothing)
+
+    checkStatusCache (cachedFileStatus, cachedKey, cachedResult) = do
+
+      -- Now we have to go probe the file system.
+      res <- probeFileSystem root cachedFileStatus
+      case res of
+        -- Some monitored file has changed
+        Left changedPath -> return (Changed (Just changedPath))
+
+        -- No monitored file has changed
+        Right (cachedFileStatus', cacheStatus) -> do
+
+          -- But we might still want to update the cache
+          whenCacheChanged cacheStatus $
+            rewriteCache cachedFileStatus' cachedKey cachedResult
+
+          let monitorFiles = reconstructMonitorFilePaths cachedFileStatus'
+          return (Unchanged (cachedResult, monitorFiles))
+
+    rewriteCache cachedFileStatus' cachedKey cachedResult = 
+      Binary.encodeFile monitorStateFile
+                        (cachedFileStatus', cachedKey, cachedResult)
+
+-- | Probe the file system to see if any of the monitored files have changed.
+--
+-- It returns Nothing if any file changed, or returns a possibly updated
+-- file 'MonitorStateFileSet' plus an indicator of whether it actually changed.
+--
+-- We may need to update the cache since there may be changes in the filesystem
+-- state which don't change any of our affected files.
+--
+-- Consider the glob @{proj1,proj2}/*.cabal@. Say we first run and find a
+-- @proj1@ directory containing @proj1.cabal@ yet no @proj2@. If we later run
+-- and find @proj2@ was created, yet contains no files matching @*.cabal@ then
+-- we want to update the cache despite no changes in our relevant file set.
+-- Specifically, we should add an mtime for this directory so we can avoid
+-- re-traversing the directory in future runs.
+--
+probeFileSystem :: FilePath -> MonitorStateFileSet
+                -> IO (Either FilePath (MonitorStateFileSet, CacheChanged))
+probeFileSystem root (MonitorStateFileSet singlePaths globPaths) =
+  runChangedM $
+    MonitorStateFileSet
+      <$> Map.traverseWithKey (probeFileStatus root)     singlePaths
+      <*> traverse            (probeGlobStatus root ".") globPaths
+
+-----------------------------------------------
+-- Monad for checking for file system changes
+--
+-- We need to be able to bail out if we detect a change (using ExceptT),
+-- but if there's no change we need to be able to rebuild the monitor
+-- state. And we want to optimise that rebuilding by keeping track if
```

Comment:
Consider an example like this:

 * matching glob: ./*/*.cabal
 * initial files: ./foo/foo.cabal  ./tmp/

So the cache has info that the top level dir has foo and tmp, and has subtrees that tell us that foo contains a match but tmp does not.

Now I do mkdir bar; touch bar/blah and re-check. There's a new dir at the top level that matches the '*' glob. So we want to check if anything underneath matches (if it does then that's an actual change). But if nothing matches then we still want to update the cache about what is there (we might not discover there's no matches 'til we've done a great deal of work). So this is a circumstance in which we will rebuild and rewrite the cache. There's potentially a lot of work that would otherwise have to be repeated. Think about unpacking a new package (but where for some reason it didn't end up matching our glob), if we didn't update the cache we might have to rescan the whole thing every time (depending on the glob pattern).

There are other cases where we could update the cache, but we don't because the what we save in those cases is more limited, and there is a cost to rewriting the cache. But the new subtree case is potentially a large amount of work that'd have to be repeated each time.